File: moves.py
from gpiozero import Servo
import evdev

class moves:
  # Front left leg
  servo0 = Servo(2, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)
  servo1 = Servo(3, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)

  # Front right leg
  servo2 = Servo(14, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)
  servo3 = Servo(15, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)

  # Back left leg
  servo4 = Servo(19, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)
  servo5 = Servo(26, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)
  
  # Back right leg
  servo6 = Servo(20, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)
  servo7 = Servo(21, min_pulse_width=0.5/1000, max_pulse_width=2.5/1000)

  devices = [evdev.InputDevice(path) for path in evdev.list_devices()]
  for device in devices:
    print(device.path, device.name, device.phys)

  device = evdev.InputDevice('/dev/input/event8')

  print(device)
  print("Base")

  servo0.value = None
  servo1.value = None

  servo2.value = None
  servo3.value = None

  servo4.value = None
  servo5.value = None

  servo6.value = None
  servo7.value = None

  while True:
    for event in device.read_loop():
      if event.type == evdev.ecodes.EV_KEY:
        # print(evdev.categorize(event))

        # Up
        if event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_X and event.value == 1:
          print("Up")

          servo0.value = 0.5
          servo1.value = 0.5

          servo2.value = 0.5
          servo3.value = 0.5

          servo4.value = 0.5
          servo5.value = 0.5

          servo6.value = 0.5
          servo7.value = 0.5

        elif event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_X and event.value == 0:
          print("Idle")

          servo0.value = None
          servo1.value = None

          servo2.value = None
          servo3.value = None

          servo4.value = None
          servo5.value = None

          servo6.value = None
          servo7.value = None


        # Left
        if event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_Y and event.value == 1:
          print("Left")

          servo0.value = -1
          servo1.value = -1

          servo2.value = -1
          servo3.value = -1

          servo4.value = -1
          servo5.value = -1

          servo6.value = -1
          servo7.value = -1

        elif event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_Y and event.value == 0:
          print("Idle")

          servo0.value = None
          servo1.value = None

          servo2.value = None
          servo3.value = None

          servo4.value = None
          servo5.value = None

          servo6.value = None
          servo7.value = None


        # Right
        if event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_B and event.value == 1:
          print("Right")

          servo0.value = 1
          servo1.value = 1

          servo2.value = -1
          servo3.value = -1

          servo4.value = -1
          servo5.value = -1

          servo6.value = -1
          servo7.value = -1

        elif event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_B and event.value == 0:
          print("Idle")

          servo0.value = None
          servo1.value = None

          servo2.value = None
          servo3.value = None

          servo4.value = None
          servo5.value = None

          servo6.value = None
          servo7.value = None


        # Down
        if event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_A and event.value == 1:
          print("Down")

          servo0.value = -1
          servo1.value = -1

          servo2.value = -1
          servo3.value = -1

          servo4.value = -1
          servo5.value = -1

          servo6.value = -1
          servo7.value = -1

        elif event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_A and event.value == 0:
          print("Idle")

          servo0.value = None
          servo1.value = None

          servo2.value = None
          servo3.value = None

          servo4.value = None
          servo5.value = None

          servo6.value = None
          servo7.value = None


          # Shutdown
        if event.type == evdev.ecodes.EV_KEY and event.code == evdev.ecodes.BTN_MODE:
          exit()

File: moves.c
#include <fcntl.h>
#include <linux/input.h>
#include <softPwm.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <wiringPi.h>

#define MIN_PULSE_WIDTH 10
#define MAX_PULSE_WIDTH 30

#define servo0 8 // GPIO 2
#define servo1 9 // GPIO 3
#define servo2 7 // GPIO 4

#define servo3 15 // GPIO 14
#define servo4 16 // GPIO 15
#define servo5 1 // GPIO 18

#define servo6 23 // GPIO 13
#define servo7 24 // GPIO 19
#define servo8 25 // GPIO 26

#define servo9 27 // GPIO 16
#define servo10 28 // GPIO 20
#define servo11 29 // GPIO 21

int main(int argc, char *argv[]) {
	if (argc != 2) {
		printf("Usage: %s <event-file-path>\n", argv[0]);
		exit(-1);
	}

	printf("Keylogger active\n");

	int fd = open(argv[1], O_RDONLY, 0);
	printf("Opened file descriptor: %d\n", fd);

	struct input_event ie;

	// 10 is 0 degrees
	// 20 is 180 degrees
	// 15 is middle (start position)
	wiringPiSetup();

	// Front left leg
	printf("Front left leg base\n");
	softPwmCreate(servo0, 15, 100); // Good
	softPwmCreate(servo1, 15, 100); // Good
	softPwmCreate(servo2, 15, 100); // Good

	// Front right leg
	printf("Front right leg base\n");
	softPwmCreate(servo3, 15, 100); // Good
	softPwmCreate(servo4, 15, 100); // Good 
	softPwmCreate(servo5, 15, 100); // Good 

	// Back left leg
	printf("Back left leg base \n");
	softPwmCreate(servo6, 15, 100); // Good 
	softPwmCreate(servo7, 15, 100); // Good
	softPwmCreate(servo8, 15, 100); // Good

	// Back right leg
	printf("Back right leg base\n");
	softPwmCreate(servo9, 15, 100); // Good
	softPwmCreate(servo10, 15, 100); // Good
	softPwmCreate(servo11, 15, 100); // Good

	while (1) {
		read(fd, &ie, sizeof(ie));

		if (ie.code == 307 && ie.value == 1) {

			printf("Front\n");

			softPwmWrite(servo0, 10);
			// delay(500);

			softPwmWrite(servo7, 10);
			softPwmWrite(servo6, 20);
			delay(500);

			softPwmWrite(servo7, 15);

			softPwmWrite(servo1, 10);
			softPwmWrite(servo0, 15);
			softPwmWrite(servo2, 22);
			delay(500);

			softPwmWrite(servo1, 15);
			delay(1000);

			softPwmWrite(servo2, 15);
			
			softPwmWrite(servo3, 20);

		} else if (ie.code == 307 && ie.value == 0) {

			printf("Idle\n");

			// Front left leg
			softPwmWrite(servo0, 15);
			softPwmWrite(servo1, 15);
			softPwmWrite(servo2, 15);

			// Front right leg
			softPwmWrite(servo3, 15);
			softPwmWrite(servo4, 15);
			softPwmWrite(servo5, 15);

			// Back left leg
			softPwmWrite(servo6, 15);
			softPwmWrite(servo7, 15);
			softPwmWrite(servo8, 15);

			// Back right leg
			softPwmWrite(servo9, 15);
			softPwmWrite(servo10, 15);
			softPwmWrite(servo11, 15);

		}





		if (ie.code == 305 && ie.value == 1) {

			printf("Turn right\n");

			// Front left leg
			softPwmWrite(servo0, 20);
			softPwmWrite(servo1, 15);
			softPwmWrite(servo2, 15);

			// Front right leg
			softPwmWrite(servo3, 20);
			softPwmWrite(servo4, 15);
			softPwmWrite(servo5, 15);

			// Back left leg
			softPwmWrite(servo6, 20);
			softPwmWrite(servo7, 15);
			softPwmWrite(servo8, 15);

			// Back right leg
			softPwmWrite(servo9, 20);
			softPwmWrite(servo10, 15);
			softPwmWrite(servo11, 15);

		} else if (ie.code == 305 && ie.value == 0) {

			printf("Idle\n");

			// Front left leg
			softPwmWrite(servo0, 15);
			softPwmWrite(servo1, 15);
			softPwmWrite(servo2, 15);

			// Front right leg
			softPwmWrite(servo3, 15);
			softPwmWrite(servo4, 15);
			softPwmWrite(servo5, 15);

			// Back left leg
			softPwmWrite(servo6, 15);
			softPwmWrite(servo7, 15);
			softPwmWrite(servo8, 15);

			// Back right leg
			softPwmWrite(servo9, 15);
			softPwmWrite(servo10, 15);
			softPwmWrite(servo11, 15);

		}





		if (ie.code == 304 && ie.value == 1) {

			printf("Back\n");

		} else if (ie.code == 304 && ie.value == 0) {

			printf("Idle\n");
			
			// Front left leg
			softPwmWrite(servo0, 15);
			softPwmWrite(servo1, 15);
			softPwmWrite(servo2, 15);
			
			// Front right leg
			softPwmWrite(servo3, 15);
			softPwmWrite(servo4, 15);
			softPwmWrite(servo5, 15);

			// Back left leg
			softPwmWrite(servo6, 15);
			softPwmWrite(servo7, 15);
			softPwmWrite(servo8, 15);

			// Back right leg
			softPwmWrite(servo9, 15);
			softPwmWrite(servo10, 15);
			softPwmWrite(servo11, 15);

		}





		if (ie.code == 308 && ie.value == 1) {

			printf("Turn left\n");

			// Front left leg
			softPwmWrite(servo0, 10);
			softPwmWrite(servo1, 15);
			softPwmWrite(servo2, 15);

			// Front right leg
			softPwmWrite(servo3, 10);
			softPwmWrite(servo4, 15);
			softPwmWrite(servo5, 15);

			// Back left leg
			softPwmWrite(servo6, 10);
			softPwmWrite(servo7, 15);
			softPwmWrite(servo8, 15);

			// Back right leg
			softPwmWrite(servo9, 10);
			softPwmWrite(servo10, 15);
			softPwmWrite(servo11, 15);

		} else if (ie.code == 308 && ie.value == 0) {

			printf("Idle\n");

			// Front left leg
			softPwmWrite(servo0, 15);
			softPwmWrite(servo1, 15);
			softPwmWrite(servo2, 15);

			// Front right leg
			softPwmWrite(servo3, 15);
			softPwmWrite(servo4, 15);
			softPwmWrite(servo5, 15);

			// Back left leg
			softPwmWrite(servo6, 15);
			softPwmWrite(servo7, 15);
			softPwmWrite(servo8, 15);

			// Back right leg
			softPwmWrite(servo9, 15);
			softPwmWrite(servo10, 15);
			softPwmWrite(servo11, 15);

		}
	}

	return 0;
}
